dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyPath.java

```diff
@@ -16,35 +16,52 @@
 
 package com.google.cloud.tools.opensource.dependencies;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
-import java.util.stream.Collectors;
 import org.eclipse.aether.artifact.Artifact;
 
-/**
- * A representation of a dependency path from a root Artifact to a node Artifact.
- */
+/** Path from the root to a node Artifact in a dependency tree. */
 public final class DependencyPath {
 
-  private List<Artifact> path = new ArrayList<>();
+  private List<Node> path = new ArrayList<>();
 
   @VisibleForTesting
-  public void add(Artifact artifact) {
-    path.add(artifact);
+  public void add(Artifact artifact, String scope, Boolean optional) {
+    path.add(new Node(artifact, scope, optional));
   }
-  
+
+  public int size() {
+    return path.size();
+  }
+
+  public Artifact getLeaf() {
+    return path.get(size() - 1).getArtifact();
+  }
+
+  public ImmutableList<Artifact> getPath() {
+    return path.stream().map(Node::getArtifact).collect(toImmutableList());
+  }
+
+  public Artifact get(int i) {
+    return path.get(i).getArtifact();
+  }
+
   @Override
   public String toString() {
-    return Joiner.on(" / ")
-        .join(path.stream().map(Artifacts::toCoordinates).collect(Collectors.toList()));
+    return Joiner.on(" / ").join(path);
   }
   
   @Override
   public boolean equals(Object o) {
-    if (o == null || !(o instanceof DependencyPath)) {
```

Comment:
o == null is already covered by `o instanceof DependencyPath`. (as per IntelliJ)